src/main/java/jenkins/plugins/logstash/LogstashWriter.java

```diff
@@ -0,0 +1,156 @@
+ * Each call to write (one line or multiple lines) sends a Logstash payload to the DAO.
+ * If any write fails, writer will not attempt to send any further messages to logstash during this build.
+ *
+ * @author Rusty Gerard
+ * @author Liam Newman
+ * @since 1.0.5
+ */
+public class LogstashWriter {
+
+  final OutputStream errorStream;
+  final AbstractBuild build;
+  final BuildData buildData;
+  final String jenkinsUrl;
+  final LogstashIndexerDao dao;
+  private boolean connectionBroken;
+
+  public LogstashWriter(AbstractBuild build, OutputStream error) {
+    this.errorStream = error != null ? error : System.err;
+    this.build = build;
+    this.dao = this.getDaoOrNull();
+    if (isConnectionBroken()) {
+      this.jenkinsUrl = null;
+      this.buildData = null;
+    } else {
+      this.jenkinsUrl = getJenkinsUrl();
+      this.buildData = getBuildData();
+    }
+
+  }
+
+  public void write(String line) {
+    if (!isConnectionBroken() && StringUtils.isNotEmpty(line)) {
+      this.write(Arrays.asList(line));
+    }
+  }
+
+  public void writeBuildLog(int maxLines) {
+    if (!isConnectionBroken()) {
+      // FIXME: build.getLog() won't have the last few lines like "Finished: SUCCESS" because this hasn't returned yet...
+      List<String> logLines;
+      try {
+        if (maxLines < 0) {
+          logLines = build.getLog(Integer.MAX_VALUE);
+        } else {
+          logLines = build.getLog(maxLines);
+        }
+      } catch (IOException e) {
+        String msg = "[logstash-plugin]: Unable to serialize log data.\n" +
+          ExceptionUtils.getStackTrace(e);
+        logErrorMessage(msg);
+
+        // Continue with error info as logstash payload
+        logLines = Arrays.asList(msg.split("\n"));
+      }
+
+      write(logLines);
+    }
+  }
+
+  public boolean isConnectionBroken() {
+    return connectionBroken;
+  }
+
+  // Method to encapsulate calls for unit-testing
+  LogstashIndexerDao getDao() throws InstantiationException {
+    LogstashInstallation.Descriptor descriptor = LogstashInstallation.getLogstashDescriptor();
+    return IndexerDaoFactory.getInstance(descriptor.type, descriptor.host, descriptor.port, descriptor.key, descriptor.username, descriptor.password);
+  }
+
+  BuildData getBuildData() {
+    return new BuildData(build, new Date());
+  }
+
+  String getJenkinsUrl() {
+    return Jenkins.getInstance().getRootUrl();
+  }
+
+  private void write(List<String> lines) {
+    JSONObject payload = dao.buildPayload(buildData, jenkinsUrl, lines);
+    try {
+      dao.push(payload.toString());
+    } catch (IOException e) {
+      String msg = "[logstash-plugin]: Failed to send log data to " + dao.getIndexerType() + ":" + dao.getDescription() + ".\n" +
+        "[logstash-plugin]: No Further logs will be sent to " + dao.getDescription() + ".\n" +
+        ExceptionUtils.getStackTrace(e);
+      logErrorMessage(msg);
+    }
+  }
+
+  private LogstashIndexerDao getDaoOrNull() {
```

Comment:
Just call this method `getDao()`

If it was public we would mention that the returned object is nullable in the javadoc.